PhysicsTools/PatUtils/plugins/L1PrefiringWeightProducer.cc

```diff
@@ -0,0 +1,470 @@
+      if (pt < 5 && !muon.isStandAloneMuon())
+        continue;
+      double prefiringprob_mu = getPrefiringRateMuon(eta, phi, pt, fluct);
+      nonPrefiringProba[fluct] *= (1. - prefiringprob_mu);
+      nonPrefiringProbaMuon[fluct] *= (1. - prefiringprob_mu);
+    }
+  }
+  //Move global prefire weights, as well as those for muons, photons, and jets, to the event
+  iEvent.emplace(nonPrefiringProbToken_, nonPrefiringProba[0]);
+  iEvent.emplace(nonPrefiringProbUpToken_, nonPrefiringProba[1]);
+  iEvent.emplace(nonPrefiringProbDownToken_, nonPrefiringProba[2]);
+
+  iEvent.emplace(nonPrefiringProbJetToken_, nonPrefiringProbaJet[0]);
+  iEvent.emplace(nonPrefiringProbJetUpToken_, nonPrefiringProbaJet[1]);
+  iEvent.emplace(nonPrefiringProbJetDownToken_, nonPrefiringProbaJet[2]);
+
+  iEvent.emplace(nonPrefiringProbPhotonToken_, nonPrefiringProbaPhoton[0]);
+  iEvent.emplace(nonPrefiringProbPhotonUpToken_, nonPrefiringProbaPhoton[1]);
+  iEvent.emplace(nonPrefiringProbPhotonDownToken_, nonPrefiringProbaPhoton[2]);
+
+  iEvent.emplace(nonPrefiringProbMuonToken_, nonPrefiringProbaMuon[0]);
+  iEvent.emplace(nonPrefiringProbMuonUpToken_, nonPrefiringProbaMuon[1]);
+  iEvent.emplace(nonPrefiringProbMuonDownToken_, nonPrefiringProbaMuon[2]);
+}
+
+double L1PrefiringWeightProducer::getPrefiringRateEcal(double eta,
+                                                       double pt,
+                                                       TH2F* h_prefmap,
+                                                       fluctuations fluctuation) const {
+  if (h_prefmap == nullptr && !skipwarnings_)
+    edm::LogWarning("L1PrefireWeightProducer") << "Prefiring map not found, setting prefiring rate to 0 " << std::endl;
+  if (h_prefmap == nullptr)
+    return 0.;
+  //Check pt is not above map overflow
+  int nbinsy = h_prefmap->GetNbinsY();
+  double maxy = h_prefmap->GetYaxis()->GetBinLowEdge(nbinsy + 1);
+  if (pt >= maxy)
+    pt = maxy - 0.01;
+  int thebin = h_prefmap->FindBin(eta, pt);
+
+  double prefrate = h_prefmap->GetBinContent(thebin);
+
+  double statuncty = h_prefmap->GetBinError(thebin);
+  double systuncty = prefiringRateSystUncEcal_ * prefrate;
+
+  if (fluctuation == up)
+    prefrate = std::min(1., prefrate + sqrt(pow(statuncty, 2) + pow(systuncty, 2)));
+  else if (fluctuation == down)
+    prefrate = std::max(0., prefrate - sqrt(pow(statuncty, 2) + pow(systuncty, 2)));
+  return prefrate;
+}
+double L1PrefiringWeightProducer::getPrefiringRateMuon(double eta,
+                                                       double phi,
+                                                       double pt,
+                                                       fluctuations fluctuation) const {
+  double prefrate;
+  double statuncty;
+  if ((dataeraMuon_.find("2016") != std::string::npos) && (eta > 1.24 && eta < 1.6) &&
+      (phi > 2.44346 && phi < 2.79253)) {
+    if (parametrizationHotSpot_ == nullptr && !skipwarnings_)
```

Comment:
Perhaps, if you already have to touch the code you can (slightly) simplify here and in the other cases below as:
```suggestion
    if (parametrizationHotSpot_ == nullptr) {
       if (!skipwarnings_) 
          edm::LogWarning("L1PrefireWeightProducer")
               "Prefiring parametrization not found, setting prefiring rate to 0 " << eta << " " << phi << std::endl;
       return 0.;
    }
```

Reply:
By the way, you already checked [in the constructor](https://github.com/cms-sw/cmssw/pull/33758/files#diff-4052fd6f31ddfc7812747e0a524c682f5344bc807fd6f55b2228e6a7bf2731f5R149) that the file with the parameterizations for the L1 muon exists: is it really needed to make all these checks once again in every event and for every muon?

Reply:
So far, if the file is missing in the constructor, nothing really happens. What certainly can be done to avoid these frequent checks is to do it at the beginning of the `produce()` method and skip the loops over the objects in case the file is not present. We would still have to check for the individual objects in the file, but that can probably also be done in the beginning and not for each object.

Reply:
I have saved the result of the checks in the constructor and the whole calculation is now skipped if inputs are missing, reducing the checks to a minimum.